src/Storages/StorageMerge.cpp

```diff
@@ -783,12 +849,119 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
 
         /// Subordinary tables could have different but convertible types, like numeric types of different width.
         /// We must return streams with structure equals to structure of Merge table.
-        convertingSourceStream(header, storage_snapshot->metadata, aliases, modified_context, *builder, processed_stage);
+        /// Besides this we add FilterTransform if it is needed to follow row level policies.
+
+        convertingSourceStream(header,
+            storage_snapshot->metadata,
+            aliases,
+            modified_context,
+            *builder,
+            processed_stage,
+            row_policy_data);
     }
 
     return builder;
 }
 
+void ReadFromMerge::RowPolicyData::init(RowPolicyFilterPtr row_policy_filter_ptr_,
+    std::shared_ptr<DB::IStorage> storage,
+    ContextPtr local_context)
+{
+    if (row_policy_filter_ptr_)
+    {
+        row_policy_filter_ptr = row_policy_filter_ptr_;
+
+        ASTPtr expr = row_policy_filter_ptr->expression;
+
+        auto storage_metadata_snapshot = storage->getInMemoryMetadataPtr();
+        auto storage_columns = storage_metadata_snapshot->getColumns();
+        auto needed_columns = storage_columns.getAllPhysical();
+
+        auto syntax_result = TreeRewriter(local_context).analyze(expr, needed_columns);
+        auto expression_analyzer = ExpressionAnalyzer{expr, syntax_result, local_context};
+
+        actions_dag = expression_analyzer.getActionsDAG(false /* add_aliases */, false /* project_result */);
+        filter_actions = std::make_shared<ExpressionActions>(actions_dag,
+            ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
```

Comment:
It seems you can extract everything you need from `filter_actions`, there is no need in `namesDifference()` function or in `RequiredSourceColumnsVisitor`:

```
        filter_input_columns = filter_actions->getRequiredColumns();

        const auto & sample_block = filter_actions->getSampleBlock();
        const auto & output_positions = filter_actions->getResultPositions();

        size_t num_result_columns = 0;

        for (size_t output_position : output_positions)
        {
            const auto & column_name = sample_block.getByPosition(output_position).name;
            if (std::find(filter_input_columns.begin(), filter_input_columns.end(), column_name) == filter_input_columns.end())
            {
                filter_column_name = column_name;
                ++num_result_columns;
            }
        }

        if (num_result_columns != 1)
            throw ...
```

Reply:
Thanks, as I recall it was the initial approach.
I recheck if it is ok.

Reply:
Hello @vitlibar , 
this code can substitute namesDifference, yes. Actually namesDifference implements exactly the same, but in slightly more elegant way.
Let me know if you have different opinion and I should switch to your code.

What I cannot understand is how this code can substitute RequiredSourceColumnsVisitor. Could you elaborate please? 
extendNames adds columns (to retrieve from storage) that are needed exclusively to evaluate row filter.

Reply:
Ok, I see, RequiredSourceColumnsVisitor is not needed indeed, filter_actions->getRequiredColumns() is enough in for extendNames